server/plugins/LFUGens.cpp

```diff
@@ -793,109 +799,281 @@ void LFGauss_Ctor(LFGauss* unit) {
+
+    unit->mPhase = phase;
+    unit->mPhaseOffset = prev_phaseOff;
+    unit->mPhaseIncrement = phaseInc;
+}
 
+void Impulse_next_ak(Impulse* unit, int inNumSamples) {
+    float* out = ZOUT(0);
     double phase = unit->mPhase;
-    double prev_phaseOffset = unit->mPhaseOffset;
-    double phaseSlope = CALCSLOPE(phaseOffset, prev_phaseOffset);
-    phase += prev_phaseOffset;
+
+    double prev_phaseInc = unit->mPhaseIncrement;
+    float* freqIn = ZIN(0);
+    float freqMul = unit->mFreqMul;
+
+    double prev_phaseOff = unit->mPhaseOffset;
+    double phaseOff = ZIN0(1);
+    double phaseOffSlope = CALCSLOPE(phaseOff, prev_phaseOff);
+    bool phaseOffChanged = phaseOffSlope != 0.f;
 
     LOOP1(
-        inNumSamples, float z; phase += phaseSlope; if (phase >= 1.f) {
-            phase -= 1.f;
-            z = 1.f;
-        } else { z = 0.f; } phase += freq;
-        ZXP(out) = z;);
+        inNumSamples, float z = Impulse_testWrapPhase(prev_phaseInc, phase); if (phaseOffChanged) {
+            phase += phaseOffSlope;
+            Impulse_testWrapPhase(prev_phaseInc, phase);
+        } double phaseInc = ZXP(freqIn) * freqMul;
+        ZXP(out) = z;
 
-    unit->mPhase = phase - phaseOffset;
-    unit->mPhaseOffset = phaseOffset;
-}
+        phase += phaseInc; prev_phaseInc = phaseInc;);
 
+    unit->mPhase = phase;
+    unit->mPhaseOffset = phaseOff;
+    unit->mPhaseIncrement = prev_phaseInc;
+}
 
-void Impulse_next_k(Impulse* unit, int inNumSamples) {
+void Impulse_next_aa(Impulse* unit, int inNumSamples) {
     float* out = ZOUT(0);
-    float freq = ZIN0(0) * unit->mFreqMul;
+    double phase = unit->mPhase;
+
+    double phaseInc = unit->mPhaseIncrement;
+    float* freqin = ZIN(0);
+    float freqmul = unit->mFreqMul;
+
+    double prev_phaseOff = unit->mPhaseOffset;
+    float* phaseOffIn = ZIN(1);
+
+    LOOP1(inNumSamples, float z = Impulse_testWrapPhase(phaseInc, phase); float phaseOff = ZXP(phaseOffIn);
+          float phaseOffInc = phaseOff - prev_phaseOff; phase += phaseOffInc; Impulse_testWrapPhase(phaseInc, phase);
+          phaseInc = ZXP(freqin) * freqmul; ZXP(out) = z;
+
+          phase += phaseInc; prev_phaseOff = phaseOff;);
 
+    unit->mPhase = phase;
+    unit->mPhaseOffset = prev_phaseOff;
+    unit->mPhaseIncrement = phaseInc;
+}
+
+void Impulse_next_ai(Impulse* unit, int inNumSamples) {
+    float* out = ZOUT(0);
     double phase = unit->mPhase;
-    LOOP1(
-        inNumSamples, float z; if (phase >= 1.f) {
-            phase -= 1.f;
-            z = 1.f;
-        } else { z = 0.f; } phase += freq;
-        ZXP(out) = z;);
+
+    double phaseInc = unit->mPhaseIncrement;
+    float* freqin = ZIN(0);
+    float freqmul = unit->mFreqMul;
+
+    LOOP1(inNumSamples, float z = Impulse_testWrapPhase(phaseInc, phase); phaseInc = ZXP(freqin) * freqmul;
+          ZXP(out) = z; phase += phaseInc;);
 
     unit->mPhase = phase;
+    unit->mPhaseIncrement = phaseInc;
 }
 
+// Note: In calc functions,
```

Comment:
This logic is sound, but I admit, I really had to scratch my head to understand it. Could this comment perhaps be expanded?

A particular sticking point is that it isn't immediately obvious why both phase and phaseOffset are initialized to the phase input. I get it now -- the phase adjustment is based on input phase - previous phase (so, it's 0 if the phase input is constant).

Users may expect to get triggers from `Impulse.ar(0, Phasor.ar(...))` (in the same way that we expect a sine wave from `SinOsc.ar(0, Phasor.ar(...))`) but in my testing, that doesn't happen. The workaround of course would be `HPZ1.ar(Phasor.ar(...)) < 0` instead of Impulse but I guess some users may be surprised by this.

Reply:
> This logic is sound, but I admit, I really had to scratch my head to understand it. Could this comment perhaps be expanded?

Agreed. I've updated the comment to read more clearly.

>A particular sticking point is that it isn't immediately obvious why both phase and
phaseOffset are initialized to the phase input.

Part of the confusion may come from the fact that the UGen argument for
_phase offset_ on the sclang side is just called `phase`. Whereas internally, `phase` is
the actual phase state value. Throughout the UGen I've renamed local vars
`phaseInc` and `phaseOff` to simply `inc` and `off`, hopefully disambiguating
it from `phase`.

>Users may expect to get triggers from Impulse.ar(0, Phasor.ar(...)) ...

Great point, I was also surprised the Impulse doesn't support driving by the
`phase`(offset). Based on your comment I've added this to the Discussion in
the Help file, as well as suggesting the `HPZ` solution for generating an
impulse train with a `Phasor`.